src/hotspot/share/opto/predicates.cpp

```diff
@@ -528,56 +560,25 @@ class DataNodesOnPathsToTargets : public StackObj {
+                                   CountedLoopNode* new_loop_node) const {
   ResourceMark rm;
   auto is_opaque_loop_node = [](const Node* node) {
     return node->is_Opaque1();
   };
   DataNodesOnPathsToTargets data_nodes_on_path_to_targets(TemplateAssertionExpressionNode::is_maybe_in_expression,
                                                           is_opaque_loop_node);
-  const Unique_Node_List& collected_nodes = data_nodes_on_path_to_targets.collect(_opaque_node);
-  DataNodeGraph data_node_graph(collected_nodes, phase);
+  Node* bool_into_opaque_node = _opaque_node->in(1);
+  const Unique_Node_List& collected_nodes = data_nodes_on_path_to_targets.collect(bool_into_opaque_node);
+  DataNodeGraph data_node_graph(collected_nodes, _phase);
   const OrigToNewHashtable& orig_to_new = data_node_graph.clone_with_opaque_loop_transform_strategy(transform_strategy,
                                                                                                     new_control);
-  assert(orig_to_new.contains(_opaque_node), "must exist");
-  Node* opaque_node_clone = *orig_to_new.get(_opaque_node);
-  return opaque_node_clone->as_OpaqueTemplateAssertionPredicate();
-}
-
-// This class is used to replace the input to OpaqueLoopStrideNode with a new node while leaving the other nodes
-// unchanged.
-class ReplaceOpaqueStrideInput : public BFSActions {
-  Node* _new_opaque_stride_input;
-  PhaseIterGVN& _igvn;
-
- public:
-  ReplaceOpaqueStrideInput(Node* new_opaque_stride_input, PhaseIterGVN& igvn)
-      : _new_opaque_stride_input(new_opaque_stride_input),
-        _igvn(igvn) {}
-  NONCOPYABLE(ReplaceOpaqueStrideInput);
-
-  void replace_for(OpaqueTemplateAssertionPredicateNode* opaque_node) {
-    DataNodeBFS bfs(*this);
-    bfs.run(opaque_node);
-  }
-
-  bool should_visit(Node* node) const override {
-    return TemplateAssertionExpressionNode::is_maybe_in_expression(node);
-  }
-
-  bool is_target_node(Node* node) const override {
-    return node->is_OpaqueLoopStride();
-  }
-
-  void target_node_action(Node* target_node) override {
-    _igvn.replace_input_of(target_node, 1, _new_opaque_stride_input);
-  }
-};
-
-// Replace the input to OpaqueLoopStrideNode with 'new_stride' and leave the other nodes unchanged.
-void TemplateAssertionExpression::replace_opaque_stride_input(Node* new_stride, PhaseIterGVN& igvn) const {
-  ReplaceOpaqueStrideInput replace_opaque_stride_input(new_stride, igvn);
-  replace_opaque_stride_input.replace_for(_opaque_node);
+  assert(orig_to_new.contains(bool_into_opaque_node), "must exist");
+  BoolNode* bool_into_opaque_node_clone = (*orig_to_new.get(bool_into_opaque_node))->as_Bool();
+  OpaqueTemplateAssertionPredicateNode* opaque_clone =
+      new OpaqueTemplateAssertionPredicateNode(bool_into_opaque_node_clone, new_loop_node);
+  _phase->C->add_template_assertion_predicate_opaque(opaque_clone);
+  _phase->register_new_node(opaque_clone, new_control);
```

Comment:
We don't clone the `OpaqueTemplateAssertionPredicateNode` anymore with `DataNodeGraph::clone_with_opaque_loop_transform_strategy` but directly here. This allows us to easily set the `new_loop_node` for it.